ACTIONX: Fix bug with matching wells + refactor

joakim-hove opened the pull request:
When evaluating an `ACTIONX` expression like:
```
ACTIONX
    WWCT > 0.75 /
/
```
the result should contain an overall thruthness[1] and the set of wells matching the condition. In this PR the bool and the set of wells is assemble together in a small class `Action::Result`. The main new functionality in this PR is that expressions like:
```
ACTIONX
    FWPR > 1000 AND  WWCT < 0.75 /
/ 
```
work correctly. Previously the first condition `FWPR > 1000` had no matching wells - and when the empty set of wells was intersected with the wells matching `WWCT < 0.75` the set of matching wells remained empty - this is fixed in this PR. In addition the PR introduces a `Action::` namespace, and even some documentation - everything is in the large third commit.

[1]: The overall truthness in an expression like this is whether *any* well satisfies the criteria.

joakim-hove commented:
jenkins build this please

joakim-hove commented:
jenkins build this please

joakim-hove commented:
I'll let this linger over the summer holiday; taking a look at this might be a healthy(?) exercise - as it should provide a gentle introduction to the ACTIONX machinery.

bska commented on src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp:
```diff
@@ -0,0 +1,137 @@
+
+
+Result& Result::operator|=(const Result& other) {
+    this->m_result |= other.m_result;
```
Bitwise 'or' is not defined for `bool`.  You need to say
```C++
this->m_result = this->m_result || other.m_result;
```

joakim-hove replied:
I believe you - of course - and will update, but the current code compiles without warning and works as intended, maybe that is just an undefined fallback to the whatever underlying integral type?

bska commented on src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp:
```diff
@@ -0,0 +1,137 @@
+}
+
+Result& Result::operator&=(const Result& other) {
+    this->m_result &= other.m_result;
```
Bitwise 'and' is not defined for `bool`.  You need to say
```C++
this->m_result = this->m_result && other.m_result;
```

atgeirr commented on opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.hpp:
```diff
@@ -0,0 +1,87 @@
+    Result(bool result_arg, const std::vector<std::string>& wells);
+    Result(bool result_arg, const WellSet& wells);
+
+    bool result() const;
```
I think the result() and scalar() members are quite confusing.

As I understand the PR description, the (or one) problem is dealing with expressions like `FWPR > 1000` which generate a result with an empty set of wells currently. The simple fix would then be to have such expressions generate a result object containing all wells. That way, a Result is "the set of wells for which the conditions hold" always.

A possible optimization would be to have a special flag for "all wells". I think that is what you have done here, but I do not understand why this state seems to be called "scalar". Furthermore I think calling wells() will not give the correct set of wells (all of them...) for such a Result.

The result() member is also confusing. I'll assume the member is supposed to be understood like this: If result() is `true`, then I contain all wells for which the conditions hold, if result() is false, then I contain all wells for which the conditions do not hold. Then again the wells() member is suspect, or at least hard to use: its meaning depends on the value of result(). I think a Result should always behave as if result() is true, if you need to negate it I would rather have a `not` operator doing that.

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp:
```diff
@@ -0,0 +1,137 @@
+        auto well_iter = this->well_set.begin();
+        while (well_iter != this->well_set.end()) {
+            if (!other.contains(*well_iter))
+                well_iter = this->well_set.erase(well_iter);
```
Fixed

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp:
```diff
@@ -0,0 +1,137 @@
+    if (this->m_scalar)
+        *this = other;
+    else {
+        for (auto& well : other.well_set)
```
Done

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp:
```diff
@@ -0,0 +1,137 @@
+    return this->m_result;
+}
+
+std::vector<std::string> Result::wells() const {
```
Yes - we could return a `const` reference to the underlying set; the reason I converted to `std::vector` was that *everyone likes a vector` - the `std::set` is slightly less standard - as I see it. No strong feelings though.

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp:
```diff
@@ -0,0 +1,137 @@
+    return *this;
+}
+
+void Result::assign(double value) {
```
Fixed

joakim-hove commented:
Thank you for the review - I dare say you have misunderstood the code at several points, but that just amounts to saying the code was less than clear. I have now updated it - personally I think it is much clearer now - hopefully you agree.

The main problem was to differentiate between a result which has no wells - because that is not relevant:
```
FPR > 100
```
and a result of *well type* - with zero matching wells:
```
WWCT > 1.25
```
that is now solved by using pointer semantics internally on the `WellSet` - where the first example will have a `nullptr`, whereas the last example will point to an empty set - that has done away with the `scalar()` uglyness completely. 

Have also implemented `operator bool()` and dropped the `result()` method, finally some code has been rewritten to enable removing the `assign(bool)` method - we don't like setters do we?

joakim-hove commented:
jenkins build this please

joakim-hove commented:
jenkins build this please

joakim-hove commented:
jenkins build this please

joakim-hove commented:
@atgeirr - seems you are back in action; cool if you give this another round.

bska commented:
jenkins build this please

joakim-hove commented:
jenkins build this please

joakim-hove commented:
OK - I'll merge this on thursday.

joakim-hove commented:
jenkins build this please
